File: 2558-take-gifts-from-the-richest-pile/2558-take-gifts-from-the-richest-pile.cpp
class Solution {
public:
    long long pickGifts(vector<int>& gifts, int k) {
        priority_queue <long long> pq;
        long long res=0;
        for(auto g: gifts)  
            pq.push(g);
        while(k--)  
            pq.push(floor(sqrt(pq.top()))), pq.pop();
        while(!pq.empty())  
            res += pq.top(), pq.pop();
        return res;
    }
};

File: 2558-take-gifts-from-the-richest-pile/2558-take-gifts-from-the-richest-pile.java
class Solution {
    public long pickGifts(int[] gifts, int k) {
        Queue<Integer> pq = new PriorityQueue<>((a, b)->b-a);
        long total = 0;
        for(int g: gifts){
            pq.offer(g);
            total += g;
        }
        for(int i=0; i<k; i++){
            int top = pq.poll();
            int newVal = (int)Math.sqrt(top);
            total -= top - newVal;
            pq.offer(newVal);
        }
        return total;
    }
}